graylog2-web-interface/src/views/components/contexts/WidgetFocusProvider.tsx

```diff
@@ -15,27 +15,63 @@
  * <http://www.mongodb.com/licensing/server-side-public-license>.
  */
 import * as React from 'react';
-import { useState, useEffect } from 'react';
+import { useState, useEffect, useCallback } from 'react';
+import { useLocation, useHistory } from 'react-router-dom';
 import PropTypes from 'prop-types';
+import URI from 'urijs';
 
 import { useStore } from 'stores/connect';
 import { WidgetStore } from 'views/stores/WidgetStore';
 import WidgetFocusContext from 'views/components/contexts/WidgetFocusContext';
 
+const _syncWithQuery = (query: string, focusedWidget: string) => {
+  const baseUri = new URI(query)
+    .removeSearch('focused');
+
+  if (focusedWidget) {
+    return baseUri.setSearch('focused', focusedWidget).toString();
+  }
+
+  return baseUri.toString();
+};
+
+const getFocusedWidgetFromSearch = (search: string) => {
+  const uri = new URI(search);
+  const { focused } = uri.search(true);
+
+  return focused;
+};
+
 const WidgetFocusProvider = ({ children }: { children: React.ReactNode }): React.ReactElement => {
-  const [focusedWidget, setFocusedWidget] = useState(undefined);
+  const { search, pathname } = useLocation();
+  const query = pathname + search;
+  const history = useHistory();
+  const [focusedWidget, setFocusedWidget] = useState<string | undefined>();
   const widgets = useStore(WidgetStore);
 
+  useEffect(() => {
+    const paramFocusedWidget = getFocusedWidgetFromSearch(search);
+
+    if (paramFocusedWidget && focusedWidget !== paramFocusedWidget) {
+      setFocusedWidget(paramFocusedWidget);
+    }
+  }, [focusedWidget, search]);
+
   useEffect(() => {
     if (focusedWidget && !widgets.has(focusedWidget)) {
       setFocusedWidget(undefined);
     }
   }, [focusedWidget, widgets]);
 
-  const updateFocus = (widgetId: string | undefined | null) => (
-    widgetId === focusedWidget
-      ? setFocusedWidget(undefined)
-      : setFocusedWidget(widgetId));
+  const updateFocus = useCallback((widgetId: string | undefined | null) => {
+    const newFocus = widgetId === focusedWidget
```

Comment:
I do not understand this part. If the `updateFocus` function is called with the widget id of the currently focused widget, `_syncWithQuery` would be called with `undefined` and would remove the `focused` query parameter. Or am I misreading something?

Maybe we should simplify the functionality of the `updateFocus` function. We could let it set:
1. the URL only (which would then be synced with the state in `WidgetContextProvider`)
2. the state only (through `setFocusedWidget`) and synchronize with the URL in `WidgetContextProvider`

Reply:
This is again related to the bug, which only @linuspahl could reproduce. I am pretty sure this part is not needed any longer since we are using a ContextProvider and not a Store anymore, but I would like to remove this when @linuspahl can test it.